django/test/runner.py

```diff
@@ -255,10 +276,22 @@ def addExpectedFailure(self, test, err):
             err = err[0], err[1], None
         self.check_picklable(test, err)
         self.events.append(('addExpectedFailure', self.test_index, err))
+        super().addExpectedFailure(test, err)
 
     def addUnexpectedSuccess(self, test):
         self.events.append(('addUnexpectedSuccess', self.test_index))
-        self.stop_if_failfast()
+        super().addUnexpectedSuccess(test)
+
+    def wasSuccessful(self):
+        """Tells whether or not this result was a success."""
+        failure_types = ('addError', 'addFailure', 'addUnexpectedSuccess',)
```

Comment:
A set literal makes a tiny bit more sense here I think: `{'addError', 'addFailure', 'addUnexpectedSuccess',}`

Reply:
Yes it does make sense and I've changed it.

Fun fact though is that a tuple is a tiny bit faster because it can be cached in the function definition, whilst the set needs to be reconstructed each time it runs:

```python
In [2]: x = 'addUnexpectedSuccess'

In [3]: %timeit x in failure_types
54.6 ns ± 0.318 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [4]: failure_types = {'addError', 'addFailure', 'addUnexpectedSuccess'}

In [5]: %timeit x in failure_types
35.6 ns ± 0.415 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [6]: def test_tuple(x):
   ...:     failure_types = ('addError', 'addFailure', 'addUnexpectedSuccess',)
   ...:     return x in failure_types
   ...:

In [7]: def test_set(x):failure_types = ('addError', 'addFailure', 'addUnexpectedSuccess',)

In [8]: def test_set(x):failure_types = ('addError', 'addFailure', 'addUnexpectedSuccess',)

In [9]: def test_set(x):failure_types = ('addError', 'addFailure', 'addUnexpectedSuccess',)

In [10]: def test_set(x):
    ...:     failure_types = {'addError', 'addFailure', 'addUnexpectedSuccess'}
    ...:     return x in failure_types
    ...:
```

In [11]: %timeit test_tuple(x)
120 ns ± 0.585 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [12]: %timeit test_set(x)
161 ns ± 0.677 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

Reply:
Oh, interesting. Is that because of mutability? What about a frozenset then? I'm not suggesting to change it here (I don't think it matters much) but you made me curious now :)

Reply:
It's an immutability optimization yes

Unfortunately frozensets don't work since they aren't a literal type. You can see here only the tuple comes through `LOAD_CONST` - one of the "constants" for the function:

```python
In [5]: dis.dis(test_tuple)
  2           0 LOAD_CONST               1 (('addError', 'addFailure', 'addUnexpectedSuccess'))
              2 STORE_FAST               1 (failure_types)

  3           4 LOAD_FAST                0 (x)
              6 LOAD_FAST                1 (failure_types)
              8 COMPARE_OP               6 (in)
             10 RETURN_VALUE

In [6]: dis.dis(test_set)
  2           0 LOAD_CONST               1 ('addError')
              2 LOAD_CONST               2 ('addFailure')
              4 LOAD_CONST               3 ('addUnexpectedSuccess')
              6 BUILD_SET                3
              8 STORE_FAST               1 (failure_types)

  3          10 LOAD_FAST                0 (x)
             12 LOAD_FAST                1 (failure_types)
             14 COMPARE_OP               6 (in)
             16 RETURN_VALUE

In [7]: def test_frozenset(x):
   ...:     failure_types = frozenset({'addError', 'addFailure', 'addUnexpectedSuccess'})
   ...:     return x in failure_types
   ...:

In [8]: dis.dis(test_frozenset)
  2           0 LOAD_GLOBAL              0 (frozenset)
              2 LOAD_CONST               1 ('addError')
              4 LOAD_CONST               2 ('addFailure')
              6 LOAD_CONST               3 ('addUnexpectedSuccess')
              8 BUILD_SET                3
             10 CALL_FUNCTION            1
             12 STORE_FAST               1 (failure_types)

  3          14 LOAD_FAST                0 (x)
             16 LOAD_FAST                1 (failure_types)
             18 COMPARE_OP               6 (in)
             20 RETURN_VALUE

```